lib/streamlit/commands/execution_control.py

```diff
@@ -54,22 +59,76 @@ def stop() -> NoReturn:  # type: ignore[misc]
+        )
+
+        # If st.rerun(scope="fragment") is called during a full script run, we raise an
+        # exception. This occurs, of course, if st.rerun(scope="fragment") is called
+        # outside of a fragment, but it somewhat surprisingly occurs if it gets called
+        # from within a fragment during a run of the full script. While this behvior may
+        # be surprising, it seems somewhat reasonable given that the correct behavior of
+        # calling st.rerun(scope="fragment") in this situation is unclear to me:
+        #   * Rerunning just the fragment immediately may cause weirdness down the line
+        #     as any part of the script that occurs after the fragment will not be
+        #     executed.
+        #   * Waiting until the full script run completes before rerunning the fragment
+        #     seems odd (even if we normally do this before running a fragment not
+        #     triggered by st.rerun()) because it defers the execution of st.rerun().
+        #   * Rerunning the full app feels incorrect as we're seemingly ignoring the
+        #     `scope` argument.
+        # With these issues and given that it seems pretty unnatural to have a
+        # fragment-scoped rerun happen during a full script run to begin with, it seems
+        # reasonable to just disallow this completely for now.
+        if not curr_queue:
+            raise StreamlitAPIException(
+                'scope="fragment" can only be specified from `@st.fragment`-decorated '
+                "functions during fragment reruns."
+            )
+
+        assert (
+            new_queue := list(
+                dropwhile(lambda x: x != ctx.current_fragment_id, curr_queue)
+            )
+        ), "Could not find current_fragment_id in fragment_id_queue. This should never happen."
+
+        return new_queue
+
+
 @gather_metrics("rerun")
-def rerun() -> NoReturn:  # type: ignore[misc]
+def rerun(  # type: ignore[misc]
+    *,  # The scope argument can only be passed via keyword.
+    scope: Literal["app", "fragment"] = "app",
+) -> NoReturn:
     """Rerun the script immediately.
 
     When ``st.rerun()`` is called, the script is halted - no more statements will
-    be run, and the script will be queued to re-run from the top.
+    be run, and the script will be queued to re-run.
+
+    Parameters
+    ----------
+    scope : Literal["app", "fragment"]
+        Specifies what part of the app should rerun. Setting scope="app" reruns the
+        full script; scope="fragment" limits the rerun to only the fragment from which
+        this function is called. If unspecified, defaults to "app".
+
+        Note that setting scope="fragment" is only valid
+            * inside of a fragment
+            * *not* during a full script run.
+        Passing this argument to ``st.rerun()`` from outside of a fragment or within a
+        fragment but when the full script is running raises a ``StreamlitAPIException``.
```

Comment:
@sfc-gh-dmatthews heads up this docstring change is probably worth reviewing / editing

Reply:
While reading this, further extending the scope parameters may be beneficial. In certain use cases, it could be helpful to reference a specific named fragment that needs to be rerun.

For example:

st.rerun(scope='result_fragment') - named fragment
or
st.rerun(scope='parent') # referring to parent or sibling fragments

For example, this would be advantageous if you have a filter control in one fragment that queues data and displays the result in another fragment. This approach would be particularly useful for larger sites with multiple fragments, especially if some fragments have slightly longer execution times.

Reply:
Thanks for the input @robingie. Something similar was brought up as a useful addition in #8511.

I think this is a feature that we may plan to add after removing the `experimental_` feature from the fragment decorator (which is already scheduled to happen next release), but it'll require some thought on our part to ensure we have the exact desired API / behavior worked out.